_posts/2021-05-31-transaction-queue-for-namecoin-core.md

```diff
@@ -0,0 +1,38 @@
+---
+layout: post
+title: "Transaction queue for Namecoin Core"
+author: yanmaani
+tags: [News]
+---
+
+Lately, work aiming to simplify the RPC API for name management has been proceeding. This is done both for the sake of improvement itself, and to make it easier to write GUIs for Namecoin.
+
+One unique element in Namecoin is the extensive use of time-dependent transactions. This is comparatively rare in Bitcoin. Either you want to send money or you don't want to send money, but the "when" is usually ASAP. However, in Namecoin, it's pretty common to, for instance, want to renew a name *when* it's close to expiring, or register a name *when* its `NAME_NEW` input has matured.
+
+Historically, this has been a bit of an annoyance. Transactions can only* be broadcast if they have a good chance of making it into the next block, and not all transactions are intended to be broadcast immediately. The way users had to deal with it, then, was to manually keep track of what transactions they wanted broadcast and when.
+
+This isn't suitable. It's bad enough for direct users, but it also makes writing auxilliary software a pain, since the obligation will fall upon it to manage a lot of state. For example, the old name management GUI had to have a series of undocumented internal hooks into the wallet database.
+
+For this reason, a transaction queue has now been added to Namecoin to explicitly manage all that state. It was written by Brandon Roberts, Jeremy Rand, and yanmaani, who is the author of this post.  The implementation is actually shockingly simple:
+
+1. Keep a list of transactions we'd like to broadcast
+2. Each block, try to broadcast everything
+3. Remove everything that made it into the mmepool
+
+The initial plan was to have a very baroque API, where users could enter transactions and have them broadcast either when a name reached a certain age or when a transaction had a certain height. However, *after already having implemented all this*, it was realized that Bitcoin's existing locking facilities could already cover our needs, and so I took it out. Wasting the effort was a bummer, but you can't be sentimental like that. In the end it made for a better API and cleaner code. Such is life.
```

Comment:
Maybe mention that the original reason for the more complex API was that the API was first implemented in Electrum-NMC, where the simplified Namecoin Core API would be a security risk (since the server would see every transaction you have in the queue)?